核心功能，to_pb 序列化bug

问题：当struct的第一个字段`int32_t test;`被注释的情况下，才可以编译通过，否则报错
```
/usr/bin/gcc -c -m64 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -std=c++20 -DNDEBUG -o build/.objs/hello/linux/x86_64/release/src/main.cpp.o src/main.cpp
error: In file included from /usr/local/include/iguana/pb_reader.hpp:3,
                 from /usr/local/include/ylt/struct_pb/struct_pb_impl.hpp:18,
                 from /usr/local/include/ylt/struct_pb.hpp:17,
                 from src/main.cpp:3:
/usr/local/include/iguana/pb_util.hpp: In instantiation of ‘void iguana::detail::append_varint_u32(Writer&, std::index_sequence<_Idx ...>) [with long unsigned int v = 130; Writer = iguana::memory_writer; long unsigned int ...I = {0}; std::index_sequence<_Idx ...> = std::integer_sequence<long unsigned int, 0>]’:
/usr/local/include/iguana/pb_util.hpp:206:23:   required from ‘void iguana::detail::serialize_varint_u32(Writer&) [with unsigned int v = 130; Writer = iguana::memory_writer]’
/usr/local/include/iguana/pb_writer.hpp:200:30:   required from ‘void iguana::detail::to_pb_impl(Type&&, uint32_t*&, Writer&) [with unsigned int key = 130; bool omit_default_val = true; Type = const std::__cxx11::basic_string<char>&; Writer = iguana::memory_writer; uint32_t = unsigned int]’
/usr/local/include/iguana/pb_writer.hpp:135:32:   required from ‘iguana::detail::to_pb_impl<0>(const input_t&, uint32_t*&, iguana::memory_writer&)::<lambda(auto:318)> [with auto:318 = std::integral_constant<long unsigned int, 15>]’
/usr/local/include/iguana/pb_util.hpp:338:22:   required from ‘constexpr void iguana::detail::for_each_n(F&&, std::index_sequence<__indices ...>) [with F = to_pb_impl<0>(const input_t&, uint32_t*&, iguana::memory_writer&)::<lambda(auto:318)>; long unsigned int ...I = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}; std::index_sequence<__indices ...> = std::integer_sequence<long unsigned int, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15>]’
/usr/local/include/iguana/pb_writer.hpp:114:15:   required from ‘void iguana::detail::to_pb_impl(Type&&, uint32_t*&, Writer&) [with unsigned int key = 0; bool omit_default_val = true; Type = const input_t&; Writer = iguana::memory_writer; uint32_t = unsigned int]’
/usr/local/include/iguana/pb_writer.hpp:482:26:   required from ‘void iguana::to_pb(const T&, Stream&) [with T = input_t; Stream = std::__cxx11::basic_string<char>; typename std::enable_if<(ylt_refletable_v<T> || is_custom_reflection_v<T>), int>::type <anonymous> = 0]’
src/main.cpp:32:18:   required from here
/usr/local/include/iguana/pb_util.hpp:200:61: error: invalid conversion from ‘uint8_t*’ {aka ‘unsigned char*’} to ‘const char*’ [-fpermissive]
  200 |   ((temp = static_cast<uint8_t>(v >> (7 * I)), writer.write(&temp, 1)), ...);
      |                                                             ^~~~~
      |                                                             |
      |                                                             uint8_t* {aka unsigned char*}
In file included from /usr/local/include/iguana/common.hpp:4,
                 from /usr/local/include/iguana/dynamic.hpp:2,
                 from /usr/local/include/ylt/struct_pb/struct_pb_impl.hpp:17:
/usr/local/include/iguana/util.hpp:200:26: note:   initializing argument 1 of ‘void iguana::memory_writer::write(const char*, std::size_t)’
  200 |   void write(const char* data, std::size_t len) {
      |              ~~~~~~~~~~~~^~~~
```

全部demo代码
``` cpp
#include <iostream>
#include <ylt/struct_pb.hpp>

struct input_t {
	int32_t test;
	int32_t test1;
	int32_t test2;
	std::string tf1;
	std::string tf2;
	std::string tf3;
	std::string tf4;
	int32_t tf5;
	int32_t tf6;
	std::string tf7;
	int32_t tf8;
	std::string tf9;
	std::string tf10;
	std::string tf11;
	std::string tf12;
	std::string tf13;
};

int main(int argc, char** argv) {

    // 模拟数据
    input_t data;
    data.tf13 = "test";

    // 序列化
    std::string ss;
    iguana::to_pb(data, ss);

    std::cout << ss << std::endl;
    std::cout << "hello world!" << std::endl;
    return 0;
}
```

--- a/iguana/detail/traits.hpp
+++ b/iguana/detail/traits.hpp
@@ -14,6 +14,12 @@
 #include <variant>
 #include <vector>
 
+#if __cplusplus > 201703L
+#if __has_include(<span>)
+#include <span>
+#endif
+#endif
+
 #include "iguana/define.h"
 
 namespace iguana {
@@ -39,6 +45,16 @@ struct is_stdstring : is_template_instant_of<std::basic_string, T> {};
 
 template <typename T>
 struct is_tuple : is_template_instant_of<std::tuple, T> {};
+
+#if __cplusplus > 201703L
+#if __has_include(<span>)
+template <typename>
+struct is_span : std::false_type {};
+
+template <typename T, size_t Extent>
+struct is_span<std::span<T, Extent>> : std::true_type {};
+#endif
+#endif
 
 template <class T>
 struct is_sequence_container

--- a/iguana/json_reader.hpp
+++ b/iguana/json_reader.hpp
@@ -265,10 +265,22 @@ IGUANA_INLINE void from_json_impl(U &value, It &&it, It &&end) {
   }
 }
 
+template <typename T>
+constexpr size_t get_array_size(const T &t) {
+#if __cplusplus > 201703L
+#if __has_include(<span>)
+  if constexpr (is_span<T>::value)
+    return t.size();
+  else
+#endif
+#endif
+    return sizeof(T) / sizeof(decltype(std::declval<T>()[0]));
+}
+
 template <typename U, typename It, std::enable_if_t<fixed_array_v<U>, int> = 0>
 IGUANA_INLINE void from_json_impl(U &value, It &&it, It &&end) {
   using T = std::remove_reference_t<U>;
-  constexpr auto n = sizeof(T) / sizeof(decltype(std::declval<T>()[0]));
+  size_t n = get_array_size(value);
   skip_ws(it, end);
 
   if constexpr (std::is_same_v<char, std::remove_reference_t<

--- a/iguana/pb_util.hpp
+++ b/iguana/pb_util.hpp
@@ -197,7 +197,9 @@ template <uint64_t v, typename Writer, size_t... I>
 IGUANA_INLINE void append_varint_u32(Writer& writer,
                                      std::index_sequence<I...>) {
   uint8_t temp = 0;
-  ((temp = static_cast<uint8_t>(v >> (7 * I)), writer.write(&temp, 1)), ...);
+  ((temp = static_cast<uint8_t>(v >> (7 * I)),
+    writer.write((const char*)&temp, 1)),
+   ...);
 }
 
 template <uint32_t v, typename Writer>

--- a/iguana/util.hpp
+++ b/iguana/util.hpp
@@ -89,7 +89,13 @@ template <typename T>
 constexpr inline bool array_v = is_array<std::remove_cvref_t<T>>::value;
 
 template <typename Type>
-constexpr inline bool fixed_array_v = c_array_v<Type> || array_v<Type>;
+constexpr inline bool fixed_array_v = c_array_v<Type> ||
+#if __cplusplus > 201703L
+#if __has_include(<span>)
+                                      is_span<Type>::value ||
+#endif
+#endif
+                                      array_v<Type>;
 
 template <typename T>
 constexpr inline bool string_view_v =

--- a/test/unit_test.cpp
+++ b/test/unit_test.cpp
@@ -95,6 +95,23 @@ TEST_CASE("test parse item array_t") {
     CHECK(test[0] == 1);
     CHECK(test[1] == -222);
   }
+#if __cplusplus > 201703L
+#if __has_include(<span>)
+  {
+    std::vector<int> v{1, 2};
+    std::span<int> span(v);
+    std::string str;
+    iguana::to_json(span, str);
+
+    std::vector<int> v1;
+    v1.resize(2);
+    std::span<int> span1(v1);
+
+    iguana::from_json(span1, str);
+    CHECK(v == v1);
+  }
+#endif
+#endif
   {
     std::string str{"[1, -222,"};
     std::array<int, 2> test;

--- a/test/unit_test.cpp
+++ b/test/unit_test.cpp
@@ -99,16 +99,16 @@ TEST_CASE("test parse item array_t") {
 #if __has_include(<span>)
   {
     std::vector<int> v{1, 2};
-    std::span<int> span(v);
+    std::span<int> span(v.data(), v.data() + 2);
     std::string str;
     iguana::to_json(span, str);
 
     std::vector<int> v1;
     v1.resize(2);
-    std::span<int> span1(v1);
+    std::span<int> span1(v1.data(), v1.data() + 2);
 
     iguana::from_json(span1, str);
-    CHECK(v == v1);
+    assert(v == v1);
   }
 #endif
 #endif

